ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,462 @@
+  middle of some process. We might want to consider opting in to this behaviour instead
+
+
+## Workflow cancellations
+
+To handle external workflow cancellations (requested by a client or another workflow) we will use a
+similar approach to activities. By default the cancellation will attempt to cancel all currently
+executing fibers. To protect critical functionality from cancellations a `workflow.shield` method
+can be used. In which case cancellation will raise after the block is executed:
+
+
+```ruby
+class MyWorkflow < Temporal::Activity
+  def execute
+    # Can get cancelled
+    workflow.execute_activity(NonCriticalActivity)
+
+    # This block will delay any cancellations until it is finished
+    async(shield: true) do
+      workflow.execute_activity(CriticalActivity)
+      workflow.execute_activity(OtherCriticalActivity)
+    end.await
+    
+    # Can get cancelled
+    workflow.execute_activity(AnotherNonCriticalActivity)
+  rescue Temporal::Error::WorkflowCancelled
+    ...
+  end
+end
+```
+
+
+## Register signal and query handlers
+
+`calculator_workflow.rb`
+
+```ruby
+class CalculatorWorkflow < Temporal::Workflow
+  # Both signal and query handlers can be defined as Symbols (when the name matches the
+  # method) or overriden when the name is different (or cannot be used as a method name)
+  signal :add
+  signal :subtract, name: 'my.signal.subtract'
+  
+  query :value # can use a getter here
+  query :get_previous_value, name: 'my.query.previous_value'
+
+  # It is preferable in this case to initialize the values before the execution to
+  # support starting a workflow with a signal
+  def initialize
+    @value = 0
+    @previous_value = 0
+  end
+
+  def execute
+    async.await_for { @value >= 42 }
+  end
+
+  private
+  
+  attr_reader :value
+
+  def add(val)
+    @previous_value = @value
+    @value += val
+  end
+
+  def subtract(val)
+    @previous_value = @value
+    @value -= val
+  end
+
+  def get_previous_value
+    @previous_value
+  end
+end
+```
+
+#### Alternative approach
+
+To provide extra flexibility in handling signals and queries (especially from external classes and
+methods) we will allow to define signals and queries from the workflow context:
+
+```ruby
+class CalculatorWorkflow < Temporal::Workflow
+  def execute
+    value = 0
+
+    workflow.on_signal(:add) do |val|
+      value += val
+    end
+
+    workflow.on_signal('my.signal.subtract') do |val|
+      value -= val
+    end
+
+    workflow.on_query(:value) do
+      value
+    end
+
+    # A condition block will pause until the enclosed statement returns true
+    async.await_for { value >= 42 }
+  end
+end
+```
+
+
+## Signal and query a workflow
+
+```ruby
+require 'calculator_workflow'
+
+connection = Temporal::Connection.connect('localhost:7233')
+client = Temporal::Client.new(connection, namespace: 'demo')
+
+handle = client.start_workflow(SumWorkflow, 1, 2, id: 'id-1', task_queue: 'demo')
```

Comment:
What about signal-to-start/start-with-signal functionality? I recall there being some design discussions of this in temporal-ruby since the optimal API was non-obvious at first

Reply:
:+1: Should specify here in proposal. What we do in most SDKs is have a separate client method, but in Python we realized this was cleaner just making the start signal and start signal args as params to the regular start call.

Reply:
I think params to the `#start_workflow` call make the most sense, especially since the signature of the method is identical otherwise. I think this is also getting into more details than the lightweight proposal was meant for :)